File: naver_mobile_injector.ts
import { chromium, devices, ElementHandle } from 'playwright';
import https from 'https';

// Helper to fetch image and convert to Base64
function fetchImageToBase64(url: string): Promise<string> {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            const chunks: Buffer[] = [];
            res.on('data', (chunk) => chunks.push(chunk));
            res.on('end', () => {
                const buffer = Buffer.concat(chunks);
                const base64 = buffer.toString('base64');
                const type = res.headers['content-type'] || 'image/png';
                resolve(`data:${type};base64,${base64}`);
            });
        }).on('error', reject);
    });
}

async function run() {
    console.log('Fetching image and converting to Base64...');
    const targetImageUrl = 'https://via.placeholder.com/750x200?text=AdMate+Perfect';
    let base64Image = '';
    try {
        base64Image = await fetchImageToBase64(targetImageUrl);
        console.log('Image converted to Base64 successfully.');
    } catch (e) {
        console.error('Failed to fetch image:', e);
        // Fallback local pixel if fetch fails
        base64Image = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==';
    }

    const browser = await chromium.launch({ headless: false });
    const context = await browser.newContext({
        ...devices['iPhone 13'],
    });
    const page = await context.newPage();

    // Listen to browser console logs
    page.on('console', msg => console.log('BROWSER LOG:', msg.text()));

    console.log('Navigating to m.naver.com...');
    await page.goto('https://m.naver.com');

    // Scroll to trigger lazy loads
    await page.mouse.wheel(0, 300);
    await page.waitForTimeout(1000);
    await page.evaluate(() => window.scrollTo(0, 0));

    console.log('Waiting for ad containers...');
    await page.waitForTimeout(3000);

    let injected = false;
    console.log('Searching for valid ad containers in Y range 150-500...');

    // Helper function
    async function attemptInjection(handle: ElementHandle<SVGElement | HTMLElement>) {
        const bbox = await handle.boundingBox();
        if (!bbox) return false;

        // Coordinate Validation: Target roughly 150 to 500
        console.log(`Checking candidate: Y=${bbox.y}, H=${bbox.height}`);
        if (bbox.y > 140 && bbox.y < 500 && bbox.height > 50) {
            console.log('--> Valid Candidate Found!');

            await handle.evaluate((el, dataUri) => {
                el.innerHTML = ''; // Clear existing
                const newImg = document.createElement('img');
                newImg.src = dataUri;

                // Force CSS Final Perfect Version
                newImg.style.setProperty('width', '100%', 'important'); // Fit container perfectly
                newImg.style.setProperty('height', 'auto', 'important');
                newImg.style.setProperty('display', 'block', 'important');
                newImg.style.setProperty('object-fit', 'contain', 'important');
                newImg.style.setProperty('border', 'none', 'important'); // Clean look

                // Ensure container reset
                el.style.setProperty('padding', '0', 'important');
                el.style.setProperty('margin', '0', 'important');
                el.style.setProperty('background', 'transparent', 'important');

                el.appendChild(newImg);
                console.log('Injected Base64 image.');
            }, base64Image);
            return true;
        }
        return false;
    }

    // Iterate selectors
    const targetSelectors = ['.main_veta', '.id_main_banner', 'div[class*="ad"]', '#veta_top'];
    for (const selector of targetSelectors) {
        if (injected) break;
        const elements = await page.$$(selector);
        for (const el of elements) {
            if (await attemptInjection(el)) {
                injected = true;
                break;
            }
        }
    }

    // Fallback divs
    if (!injected) {
        console.log('Specific targeting failed. Scanning ALL divs in range...');
        const divs = await page.$$('div');
        for (const div of divs) {
            if (injected) break;
            const bbox = await div.boundingBox();
            // Slightly relaxed range for fallback to ensure we catch Y=180
            if (bbox && bbox.y > 150 && bbox.y < 400 && bbox.height > 50 && bbox.width > 300) {
                console.log(`Fallback Candidate: ${bbox.y}`);
                if (await attemptInjection(div)) {
                    injected = true;
                    console.log('Success with fallback div');
                }
            }
        }
    }

    console.log('Capturing screenshot...');
    await page.waitForTimeout(2000); // Wait for render
    await page.screenshot({ path: 'naver_preview_final_perfect.png', fullPage: false });
    console.log('Screenshot saved to naver_preview_final_perfect.png');

    await browser.close();
}

run().catch(err => {
    console.error(err);
    process.exit(1);
});

File: naver_mobile_injector.py
import asyncio
from playwright.async_api import async_playwright

async def run():
    async with async_playwright() as p:
        # Emulate iPhone 13
        device = p.devices['iPhone 13']
        browser = await p.chromium.launch(headless=False) # Headless=False to see it happening
        context = await browser.new_context(**device)
        page = await context.new_page()

        print("Navigating to m.naver.com...")
        await page.goto("https://m.naver.com")

        # Wait for the ad to load
        print("Waiting for ad element...")
        # Selector based on the analysis: Anchor looking for g.tivan.naver.com
        # We use a broad selector to catch the ad link.
        ad_selector = 'a[href*="g.tivan.naver.com/gfa/"] img'
        
        try:
            await page.wait_for_selector(ad_selector, timeout=10000)
            print("Ad element found!")
        except Exception as e:
            print(f"Ad element not found with standard selector. Dumping page content for debugging... {e}")
            # Fallback or just try to find any image if specific ad missing?
            # m.naver.com is dynamic, sometimes ads map differently. 
            # Let's try a more generic approach if that fails, finding an element with 'ad' text or class?
            # For now, let's proceed assuming it works or we catch the error.
        
        # Inject the image
        print("Injecting custom ad image...")
        image_url = "https://via.placeholder.com/750x200?text=AdMate+Test"
        
        # Javascript to replace the image source and style
        # We target the parent anchor to make sure we control the content, or just the img.
        # Let's find the img and change src.
        
        injection_code = f"""
        const adImage = document.querySelector('{ad_selector}');
        if (adImage) {{
            adImage.src = '{image_url}';
            adImage.srcset = '{image_url}'; // Clear srcset if present
            adImage.style.width = '100%';
            adImage.style.height = 'auto'; // Ensure aspect ratio
            adImage.style.objectFit = 'contain';
            
            // Optional: visual border to confirm injection
            adImage.style.border = '2px solid red';
            console.log('Ad injected successfully');
        }} else {{
            console.error('Ad element not found during injection');
        }}
        """
        
        await page.evaluate(injection_code)
        
        # Wait a bit for the rendering update
        await page.wait_for_timeout(2000)
        
        # Capture screenshot
        print("Capturing screenshot...")
        await page.screenshot(path="naver_preview_test.png", full_page=False)
        print("Screenshot saved to naver_preview_test.png")

        await browser.close()

if __name__ == "__main__":
    asyncio.run(run())
